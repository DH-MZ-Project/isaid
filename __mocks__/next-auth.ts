export const getServerSession = jest.fn(() =>
  Promise.resolve({
    user: {
      id: '5', // 테스트용 유저 ID
      email: 'test@test.com',
    },
  })
);
