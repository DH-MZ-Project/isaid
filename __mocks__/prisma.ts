import {
  createChallengePrismaMock,
  createEtfTestPrismaMock,
  createPrismaMock,
} from './prisma-factory';

// Jest에서 사용할 수 있도록 전역 mock 인스턴스 생성
let mockPrismaInstance = createPrismaMock();

export const prisma = mockPrismaInstance;

// 테스트에서 mock을 재설정할 수 있는 헬퍼 함수
export const resetPrismaMock = () => {
  mockPrismaInstance = createPrismaMock();
  return mockPrismaInstance;
};

// 특화된 mock으로 재설정하는 함수들
export const resetWithEtfTestPrismaMock = () => {
  mockPrismaInstance = createEtfTestPrismaMock();
  return mockPrismaInstance;
};

export const resetWithChallengePrismaMock = () => {
  mockPrismaInstance = createChallengePrismaMock();
  return mockPrismaInstance;
};

// 커스텀 overrides로 재설정
export const resetWithCustomMock = (overrides = {}) => {
  mockPrismaInstance = createPrismaMock(overrides);
  return mockPrismaInstance;
};

export const getPrismaMock = () => mockPrismaInstance;
