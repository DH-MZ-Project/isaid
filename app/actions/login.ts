'use server';

import { redirect } from 'next/navigation';

export async function login(_prev: { message?: string }, formData: FormData) {
  const email = formData.get('email')?.toString() ?? '';
  const password = formData.get('password')?.toString() ?? '';

  const ok = email === 'test@example.com' && password === '1234';

  if (!ok) return { message: '이메일이나 비밀번호가 일치하지 않습니다.' };

  redirect('/');
}
