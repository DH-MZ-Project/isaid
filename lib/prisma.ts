import { PrismaClient } from '@prisma/client';

const globalForPrisma = globalThis as unknown as {
  prisma: PrismaClient | undefined;
};

export const prisma =
  globalForPrisma.prisma ??
  new PrismaClient({
    log: ['query', 'error'],
  });

if (process.env.NODE_ENV !== 'production') {
  globalForPrisma.prisma = prisma;
}

// PrismaTransaction 타입 정의
export type PrismaTransaction = Omit<
  PrismaClient,
  '$connect' | '$disconnect' | '$on' | '$transaction' | '$use' | '$extends'
>;
