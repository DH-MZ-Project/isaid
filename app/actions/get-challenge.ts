// app/actions/getChallenges.ts
'use server';

import { getServerSession } from 'next-auth';
import dayjs from 'dayjs';
import { authOptions } from '@/lib/auth-options';
import { prisma } from '@/lib/prisma';

export type ChallengeStatus = 'CLAIMED' | 'ACHIEVABLE' | 'INCOMPLETE';

export type ChallengeInfo = {
  id: string;
  issueName: string;
  title: string;
  challengeDescription: string;
  quantity: number;
  status: ChallengeStatus;
};

export async function getChallenges(): Promise<ChallengeInfo[]> {
  const session = await getServerSession(authOptions);
  if (!session?.user?.id) throw new Error('Not authenticated');
  const userId = BigInt(session.user.id);
  const today = dayjs().startOf('day');

  const rows = await prisma.challenge.findMany({
    include: {
      etf: { select: { issueName: true } },
      userChallengeClaims: {
        where: { userId },
        select: { claimDate: true },
      },
      userChallengeProgresses: {
        where: { userId },
        select: { progressVal: true },
      },
    },
  });

  return rows.map((c) => {
    const hasClaim = c.userChallengeClaims.length > 0;
    const progress = c.userChallengeProgresses[0]?.progressVal ?? 0;

    let status: ChallengeStatus;
    switch (c.challengeType) {
      case 'ONCE':
        status = hasClaim
          ? 'CLAIMED'
          : progress >= 1
            ? 'ACHIEVABLE'
            : 'INCOMPLETE';
        break;
      case 'STREAK':
        status = hasClaim
          ? 'CLAIMED'
          : progress >= 7
            ? 'ACHIEVABLE'
            : 'INCOMPLETE';
        break;
      case 'DAILY': {
        const claimedToday = c.userChallengeClaims.some((cl) =>
          today.isSame(dayjs(cl.claimDate), 'day')
        );
        status = claimedToday
          ? 'CLAIMED'
          : progress > 0
            ? 'ACHIEVABLE'
            : 'INCOMPLETE';
        break;
      }
      default:
        status = 'INCOMPLETE';
    }

    return {
      id: c.id.toString(),
      issueName: c.etf.issueName ?? '',
      title: c.title,
      challengeDescription: c.challengeDescription,
      quantity: Number(c.quantity),
      status,
    };
  });
}
