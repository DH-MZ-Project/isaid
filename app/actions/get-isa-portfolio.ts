import { getServerSession } from 'next-auth';
import { authOptions } from '@/lib/auth-options';
import { prisma } from '@/lib/prisma';

export async function getISAPortfolio(yearMonth: string) {
  const session = await getServerSession(authOptions);

  if (!session || !session.user?.id) {
    throw new Error('Unauthorized');
  }

  const [year, month] = yearMonth.split('-').map(Number);
  const snapshotDate = new Date(Date.UTC(year, month, 0));

  const isa = await prisma.iSAAccount.findUnique({
    where: { userId: BigInt(session.user.id) },
    select: {
      generalHoldingSnapshots: {
        where: {
          snapshotDate: {
            gte: snapshotDate,
            lt: new Date(snapshotDate.getTime() + 24 * 60 * 60 * 1000),
          },
        },
        select: {
          evaluatedAmount: true,
          snapshotType: true,
        },
      },
      generalHoldings: {
        select: {
          totalCost: true,
          product: {
            select: {
              instrumentType: true,
            },
          },
        },
      },
      etfHoldingSnapshots: {
        where: {
          snapshotDate: {
            gte: snapshotDate,
            lt: new Date(snapshotDate.getTime() + 24 * 60 * 60 * 1000),
          },
        },
        select: {
          evaluatedAmount: true,
          etf: {
            select: {
              idxMarketType: true, // "국내", "해외", "국내&해외"
            },
          },
        },
      },
    },
  });

  if (!isa) throw new Error('ISA account not found');

  console.log(`Portfolio snapshot for ${yearMonth}`);

  let bond = 0;
  let fund = 0;
  let els = 0;

  for (const holding of isa.generalHoldings) {
    if (holding.product?.instrumentType === 'BOND') {
      bond += Number(holding.totalCost);
    } else if (holding.product?.instrumentType === 'FUND') {
      fund += Number(holding.totalCost);
    } else {
      els += Number(holding.totalCost);
    }
  }

  let etfDomestic = 0;
  let etfForeign = 0;
  let etfBoth = 0;
  for (const etf of isa.etfHoldingSnapshots) {
    const value = Number(etf.evaluatedAmount);
    const type = etf.etf.idxMarketType;
    if (type === '국내') etfDomestic += value;
    else if (type === '해외') etfForeign += value;
    else if (type === '국내&해외') etfBoth += value;
  }

  console.log('ETF Domestic total:', etfDomestic);
  console.log('ETF Foreign total:', etfForeign);
  console.log('ETF Domestic&Foreign total:', etfBoth);

  const total = bond + fund + els + etfDomestic + etfForeign + etfBoth;

  const format = (label: string, value: number) => ({
    category: label,
    value,
    percentage: total > 0 ? Number(((value / total) * 100).toFixed(1)) : 0,
  });

  return [
    format('채권', bond),
    format('펀드', fund),
    format('ELS', els),
    format('국내 ETF', etfDomestic),
    format('해외 ETF', etfForeign),
    format('국내&해외 ETF', etfBoth),
  ];
}
