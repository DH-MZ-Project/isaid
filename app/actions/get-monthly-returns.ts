'use server';

import { getServerSession } from 'next-auth';
import { authOptions } from '@/lib/auth-options';
import { prisma } from '@/lib/prisma';

const MonthDate = {
  '1': {
    value: '2025-01-31',
  },
  '2': {
    value: '2025-02-28',
  },
  '3': {
    value: '2025-03-31',
  },
  '4': {
    value: '2025-04-30',
  },
  '5': {
    value: '2025-05-30',
  },
  '6': {
    value: '2025-06-30',
  },
} as const;

type MonthKey = keyof typeof MonthDate;

export async function getMonthlyReturns(month: MonthKey) {
  try {
    const session = await getServerSession(authOptions);
    if (!session?.user?.id) throw new Error('로그인이 필요합니다.');
    const userId = Number(session.user.id);

    const isaAccount = await prisma.iSAAccount.findUnique({
      where: { userId },
      select: { id: true, accountType: true },
    });

    const isaAccountId = Number(isaAccount?.id);
    if (!isaAccountId) {
      throw new Error('ISA 계좌가 없습니다.');
    }
    //월별 수익률 목록
    const returns = await prisma.monthlyReturn.findMany({
      where: {
        isaAccountId: isaAccountId,
      },
      orderBy: {
        baseDate: 'asc',
      },
      select: {
        baseDate: true,
        entireProfit: true,
      },
    });

    const formattedReturns = returns.map((r) => ({
      [r.baseDate.toISOString().slice(0, 10)]: Number(
        (Number(r.entireProfit) * 100).toFixed(2)
      ),
    }));
    const baseDate = MonthDate[month].value;

    // 실제 데이터베이스에 저장된 시간을 고려하여 범위 설정
    // 데이터가 2025-06-30T23:59:59.000Z에 저장되어 있음
    const startOfDay = new Date(`${baseDate}T00:00:00.000Z`);
    const endOfDay = new Date(`${baseDate}T23:59:59.999Z`);

    // 평가금액 계산= etf+general+ 현금
    const etfSnapshots = await prisma.eTFHoldingSnapshot.findMany({
      where: {
        isaAccountId,
        snapshotDate: {
          gte: startOfDay,
          lte: endOfDay,
        },
      },
      select: {
        etfId: true,
        evaluatedAmount: true,
        snapshotDate: true,
      },
    });

    const etfEvaluatedAmount = etfSnapshots.reduce((sum, snap) => {
      return sum + Number(snap.evaluatedAmount ?? 0);
    }, 0);

    // General 데이터 조회 전에 전체 데이터 확인
    const allGeneralSnapshots = await prisma.generalHoldingSnapshot.findMany({
      where: {
        isaAccountId,
      },
      select: {
        snapshotDate: true,
        snapshotType: true,
        evaluatedAmount: true,
      },
      orderBy: {
        snapshotDate: 'desc',
      },
      take: 10,
    });

    const generalEvaluated = await prisma.generalHoldingSnapshot.aggregate({
      where: {
        isaAccountId,
        snapshotDate: {
          gte: startOfDay,
          lte: endOfDay,
        },
        snapshotType: {
          in: ['GENERAL', 'CASH'],
        },
      },
      _sum: {
        evaluatedAmount: true,
      },
    });

    // snapshotType 필터 없이 조회해보기
    const generalWithoutTypeFilter =
      await prisma.generalHoldingSnapshot.aggregate({
        where: {
          isaAccountId,
          snapshotDate: {
            gte: startOfDay,
            lte: endOfDay,
          },
        },
        _sum: {
          evaluatedAmount: true,
        },
      });

    const totalEvaluatedAmount =
      Number(etfEvaluatedAmount ?? 0) +
      Number(generalEvaluated._sum.evaluatedAmount ?? 0);

    const totalInvestedAmount = 17_000_000; // 고정 초기 투자금

    const evaluatedProfit = totalEvaluatedAmount - totalInvestedAmount;

    return {
      returns: formattedReturns,
      evaluatedAmount: totalEvaluatedAmount,
      evaluatedProfit: evaluatedProfit,
    };
  } catch (error: unknown) {
    if (error instanceof Error) {
      console.error('[getMonthlyReturns] Error:', error.message);
    } else {
      console.error('[getMonthlyReturns] Unknown error:', error);
    }
    throw new Error('Failed to fetch monthly returns');
  }
}
