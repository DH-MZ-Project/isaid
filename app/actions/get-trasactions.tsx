import { getServerSession } from 'next-auth';
import { authOptions } from '@/lib/auth-options';
import { prisma } from '@/lib/prisma';

export type CalendarTx = {
  title: string;
  amount: number;
  type: '매수' | '매도' | '입금' | '출금' | '배당금' | '이자';
};

export type CalendarData = {
  transactionDates: Date[];
  transactionData: Record<string, CalendarTx[]>;
};

export const getTransactions = async (): Promise<CalendarData> => {
  const session = await getServerSession(authOptions);
  if (!session?.user?.id) throw new Error('로그인이 필요합니다.');
  const userId = Number(session.user.id);

  const isaAccount = await prisma.iSAAccount.findUnique({
    where: { userId },
    select: { id: true },
  });
  if (!isaAccount) throw new Error('ISA 계좌가 없습니다.');
  const isaAccountId = isaAccount.id;

  const [generalTxs, etfTxs] = await Promise.all([
    prisma.generalTransaction.findMany({
      where: { isaAccountId },
      select: {
        id: true,
        transactionType: true,
        price: true,
        transactionAt: true,
        product: { select: { productName: true } },
      },
    }),
    prisma.eTFTransaction.findMany({
      where: { isaAccountId },
      select: {
        id: true,
        transactionType: true,
        price: true,
        quantity: true,
        transactionAt: true,
        etf: { select: { issueNameKo: true } },
      },
    }),
  ]);

  type TxCommon = {
    id: string;
    rawType: string;
    amount: number;
    at: Date;
    label: string;
  };
  const normalized: TxCommon[] = [
    ...generalTxs.map((g) => ({
      id: g.id.toString(),
      rawType: g.transactionType,
      amount: g.price?.toNumber() ?? 0,
      at: g.transactionAt,
      label: g.product.productName,
    })),
    ...etfTxs.map((e) => ({
      id: e.id.toString(),
      rawType: e.transactionType,
      amount: e.price.toNumber(),
      at: e.transactionAt,
      label: e.etf.issueNameKo ?? 'ETF',
    })),
  ];

  const typeMap: Record<string, CalendarTx['type']> = {
    BUY: '매수',
    SELL: '매도',
    DIVIDEND: '배당금',
    INTEREST: '이자',
    DEPOSIT: '입금',
    WITHDRAW: '출금',
  };

  const transactionData: Record<string, CalendarTx[]> = {};
  normalized.forEach((tx) => {
    const key = tx.at.toISOString().slice(0, 10);
    if (!transactionData[key]) transactionData[key] = [];
    transactionData[key].push({
      title: tx.label,
      amount: tx.amount,
      type: typeMap[tx.rawType] ?? '매수',
    });
  });

  const transactionDates = Object.keys(transactionData).map((dateStr) => {
    const [y, m, d] = dateStr.split('-').map(Number);
    return new Date(y, m - 1, d);
  });

  return {
    transactionDates,
    transactionData,
  };
};
