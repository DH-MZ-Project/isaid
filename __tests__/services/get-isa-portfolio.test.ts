import { getServerSession } from 'next-auth';
import { getISAPortfolio } from '@/app/actions/get-isa-portfolio';
import { prisma } from '@/lib/prisma';

// next-auth의 getServerSession 함수 모킹
jest.mock('next-auth');
// prisma 클라이언트의 iSAAccount 모델 모킹
jest.mock('@/lib/prisma', () => ({
  prisma: {
    iSAAccount: {
      findUnique: jest.fn(),
    },
  },
}));

describe('getISAPortfolio', () => {
  // 각 테스트 전 모킹된 함수들의 호출 기록 초기화
  beforeEach(() => {
    jest.clearAllMocks();
  });

  // 세션이 없을 경우 'Unauthorized' 에러가 발생하는지 테스트
  it('throws if no session is found', async () => {
    (getServerSession as jest.Mock).mockResolvedValue(null);
    await expect(getISAPortfolio('2025-06')).rejects.toThrow('Unauthorized');
  });

  // ISA 계좌에서 각 자산의 비율이 정확히 계산되는지 검증
  it('returns correct portfolio breakdown', async () => {
    // 로그인된 유저 세션 모킹
    (getServerSession as jest.Mock).mockResolvedValue({
      user: { id: '1' },
    });

    // prisma iSAAccount.findUnique 함수가 반환할 모킹 데이터 설정
    (prisma.iSAAccount.findUnique as jest.Mock).mockResolvedValue({
      generalHoldingSnapshots: [],
      generalHoldings: [
        {
          totalCost: 1000000,
          product: { instrumentType: 'BOND' },
        },
        {
          totalCost: 2000000,
          product: { instrumentType: 'FUND' },
        },
        {
          totalCost: 3000000,
          product: { instrumentType: 'ELS' },
        },
      ],
      etfHoldingSnapshots: [
        {
          evaluatedAmount: 4000000,
          etf: { idxMarketType: '국내' },
        },
        {
          evaluatedAmount: 5000000,
          etf: { idxMarketType: '해외' },
        },
        {
          evaluatedAmount: 6000000,
          etf: { idxMarketType: '국내&해외' },
        },
      ],
    });

    // 실제 함수 호출 및 결과 저장
    const result = await getISAPortfolio('2025-06');

    // 반환된 포트폴리오가 예상한 카테고리별 금액과 비율로 정확히 계산되었는지 검증
    expect(result).toEqual([
      { category: '채권', value: 1000000, percentage: 4.8 },
      { category: '펀드', value: 2000000, percentage: 9.5 },
      { category: 'ELS', value: 3000000, percentage: 14.3 },
      { category: '국내 ETF', value: 4000000, percentage: 19.0 },
      { category: '해외 ETF', value: 5000000, percentage: 23.8 },
      { category: '국내&해외 ETF', value: 6000000, percentage: 28.6 },
    ]);
  });

  // 반환된 포트폴리오의 전체 구조가 스냅샷과 일치하는지 확인
  it('matches snapshot for consistent portfolio structure', async () => {
    (getServerSession as jest.Mock).mockResolvedValue({ user: { id: '1' } });

    (prisma.iSAAccount.findUnique as jest.Mock).mockResolvedValue({
      generalHoldingSnapshots: [],
      generalHoldings: [
        {
          totalCost: 1000000,
          product: { instrumentType: 'BOND' },
        },
        {
          totalCost: 2000000,
          product: { instrumentType: 'FUND' },
        },
        {
          totalCost: 3000000,
          product: { instrumentType: 'ELS' },
        },
      ],
      etfHoldingSnapshots: [
        {
          evaluatedAmount: 4000000,
          etf: { idxMarketType: '국내' },
        },
        {
          evaluatedAmount: 5000000,
          etf: { idxMarketType: '해외' },
        },
        {
          evaluatedAmount: 6000000,
          etf: { idxMarketType: '국내&해외' },
        },
      ],
    });

    const result = await getISAPortfolio('2025-06');
    expect(result).toMatchSnapshot();
  });
});
