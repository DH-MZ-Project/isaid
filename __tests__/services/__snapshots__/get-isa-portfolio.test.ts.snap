// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`getISAPortfolio matches snapshot for consistent portfolio structure 1`] = `
[
  {
    "category": "채권",
    "percentage": 4.8,
    "value": 1000000,
  },
  {
    "category": "펀드",
    "percentage": 9.5,
    "value": 2000000,
  },
  {
    "category": "ELS",
    "percentage": 14.3,
    "value": 3000000,
  },
  {
    "category": "국내 ETF",
    "percentage": 19,
    "value": 4000000,
  },
  {
    "category": "해외 ETF",
    "percentage": 23.8,
    "value": 5000000,
  },
  {
    "category": "국내&해외 ETF",
    "percentage": 28.6,
    "value": 6000000,
  },
]
`;
